google-cloud-spanner/src/test/java/com/google/cloud/spanner/SessionPoolTest.java

```diff
@@ -1583,16 +1585,45 @@ public void testSessionMetrics() {
     Session session2 = pool.getReadSession();
 
     MetricsRecord record = metricRegistry.pollRecord();
+    assertThat(record.getMetrics().size()).isEqualTo(4);
     assertThat(record.getMetrics()).containsEntry(MetricRegistryConstants.IN_USE_SESSIONS, 2L);
     assertThat(record.getMetrics()).containsEntry(MetricRegistryConstants.MAX_IN_USE_SESSIONS, 2L);
+    assertThat(record.getMetrics()).containsEntry(MetricRegistryConstants.SESSIONS_TIMEOUT, 0L);
     assertThat(record.getMetrics())
         .containsEntry(
             MetricRegistryConstants.MAX_ALLOWED_SESSIONS, (long) options.getMaxSessions());
     assertThat(record.getLabels()).containsEntry(SPANNER_LABEL_KEYS, labelValues);
 
+    final CountDownLatch latch = new CountDownLatch(1);
+    // Try asynchronously to take another session. This attempt should time out.
+    Future<Void> fut =
+        executor.submit(
+            new Callable<Void>() {
+              @Override
+              public Void call() {
+                latch.countDown();
+                Session session = pool.getReadSession();
+                session.close();
+                return null;
+              }
+            });
+    // Wait until the background thread is actually waiting for a session.
+    latch.await();
+    // Wait until the request has timed out.
+    int waitCount = 0;
+    while (pool.getNumWaiterTimeouts() == 0L && waitCount < 1000) {
+      Thread.sleep(5L);
+      waitCount++;
+    }
+    // Return the checked out session to the pool so the async request will get a session and
+    // finish.
     session2.close();
-    session1.close();
+    // Verify that the async request also succeeds.
+    fut.get(10L, TimeUnit.SECONDS);
+    executor.shutdown();
 
+    session1.close();
+    assertThat(record.getMetrics()).containsEntry(MetricRegistryConstants.SESSIONS_TIMEOUT, 1L);
```

Comment:
I'm a little bit worried that this assert could turn out to be flaky. If I understand the test code correctly, the get session timeout will happen for the first time after 20ms. You poll the number of timeouts every 5ms and break the loop if you have at least 1 timeout. If the following lines of code happen to take more than 15ms to execute, you might have more than 1 timeouts in your metrics. Would it be possible to make this assert check that the number of timeouts is at least 1 (instead of equal to 1)?

Reply:
I was thinking about it, but tried multiple times locally didn't see the issue. Having said that, I understand your concern, made the change(d1c13d84d35a2e10f4620941649d022d83a8a376) to the test. PTAL when you get a chance. Thanks :)